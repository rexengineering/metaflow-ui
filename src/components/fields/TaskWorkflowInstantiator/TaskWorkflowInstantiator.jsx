import React from "react";
import {Chip, makeStyles} from "@material-ui/core";
import {useDispatch, useSelector} from "react-redux";
import PropTypes from "prop-types";
import {startWorkflowByName} from "../../../store/thunks/thunks";
import {selectActiveWorkflows, selectIsFlexTaskActive} from "../../../store/selectors/rexflow";
import isTalkTrackDidInitialized from "../../../utils/talkTracks";

const useStyles = makeStyles(({ spacing }) => ({
    chip: {
        float: "left",
        marginRight: spacing(1),
    }
}))

function TaskWorkflowInstantiator({onClick, data: workflowName, label, ...props}){
    const classes = useStyles();
    const dispatch = useDispatch();
    const activeWorkflows = useSelector(selectActiveWorkflows);
    const isInitialized =  isTalkTrackDidInitialized(activeWorkflows, workflowName);
    const isFlexTaskActive = useSelector(selectIsFlexTaskActive);

    const initializeWorkflow = () => {
        if (!workflowName || isInitialized || !isFlexTaskActive)
            return
        dispatch(startWorkflowByName(workflowName));
        onClick();
    };

    return (
        <Chip
            disabled={isInitialized || !isFlexTaskActive}
            className={classes.chip}
            onClick={initializeWorkflow}
            label={label}
            {...props}
        />
    )
}


TaskWorkflowInstantiator.defaultProps = {
    onClick: () => {},
};

TaskWorkflowInstantiator.propTypes = {
    onClick: PropTypes.func,
    label: PropTypes.string.isRequired,
    data: PropTypes.string.isRequired,
};

export default TaskWorkflowInstantiator;