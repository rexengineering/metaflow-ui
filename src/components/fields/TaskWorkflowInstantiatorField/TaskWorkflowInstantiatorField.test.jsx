import { render, screen, fireEvent } from "@testing-library/react";
import React from "react";
import TaskWorkflowInstantiator from "./TaskWorkflowInstantiatorField";
import {Provider} from "react-redux";
import getStore from "../../../store";

describe("<TaskWorkflowInstantiator />", () => {

    const workflowInstantiatorButtonProps = {
        label: "Buying",
        onClick: () => {},
        workflowIdentifier: "buying-123",
        data: "some data",
    };

    const renderTaskWorkflowInstantiator = ( {label, onClick, workflowIdentifier, data } = workflowInstantiatorButtonProps ) => (
        <Provider store={getStore({})}>
            <TaskWorkflowInstantiator data={data} label={label} onClick={onClick} workflowIdentifier={workflowIdentifier} />
        </Provider>
    );

    it("should render", () => {
        const { asFragment } = render(renderTaskWorkflowInstantiator());
        const initialRender = asFragment();

        expect(initialRender).not.toBeNull();
    });

    it("should match snapshot", () => {
        const { asFragment } = render(renderTaskWorkflowInstantiator());
        const initialRender = asFragment();

        expect(initialRender).toMatchSnapshot();
    });

    it("should render a provided label", () => {
        const customLabel = "Selling";
        render(renderTaskWorkflowInstantiator({...workflowInstantiatorButtonProps, label: customLabel}));
        const element = screen.getByText(customLabel);

        expect(element).toBeInTheDocument();
    });
});