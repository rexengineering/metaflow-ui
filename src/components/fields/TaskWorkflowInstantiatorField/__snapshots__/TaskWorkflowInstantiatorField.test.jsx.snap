// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TaskWorkflowInstantiator /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    aria-disabled="true"
    class="MuiButtonBase-root MuiChip-root makeStyles-chip-2 Mui-disabled MuiChip-clickable"
    role="button"
    tabindex="0"
  >
    <span
      class="MuiChip-label"
    >
      Buying
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </div>
</DocumentFragment>
`;
